### ISSUE TITLE
Can't import socket.io into Stencil app

### ISSUE THREAD

**Issue by bitflower** [ionic-team/stencil#495]
**Repo with the problem outlined:**
[https://github.com/bitflower/stencil-socketio-problem](https://github.com/bitflower/stencil-socketio-problem)

**Current behavior:**
Neither socket.io nor socket.io-client can be imported into Stencil.

**Steps to reproduce:**
See repo

... and then ```npm run dev```

Hope you have a minute to check this out as I'm far from being an expert in rollup! :-)

**Comment by jthoms1** [comment]
Looks like this issue was reported with rollup here.  https://github.com/rollup/rollup-plugin-commonjs/issues/61

I tried the change outlined and it appears to work.

**Comment by bitflower** [comment]
Hi @cjorasch,

I have it running for a couple of weeks with this code (Typescript):

```javascript
// Socket.IO client
import io from 'socket.io-client';

const ioFunc: any = (io as any).default ? (io as any).default : io;
const socket: any = ioFunc('http://localhost:3030', {
  transports: ['websocket'],
  forceNew: true
});
```

**Comment by kirillgroshkov** [comment]
+1

**Comment by kirillgroshkov** [comment]
Sadly it's blocking us to continue using Stencil for our projects...

**Comment by bitflower** [comment]
Check this issue. The fd erro relates to th ‚debug‘ module: https://github.com/ionic-team/stencil/issues/494

As a heavy workaround I habe exchanged all node.js files of each debug copy in node_modules with browser.js.

Ugly I know.

But my gut feeling and trust in Stencil is just too good ;-)

**Comment by kirillgroshkov** [comment]
I came up with a workaround to load `socketio-client` dynamically (not to bundle it at all), like this: https://stackoverflow.com/questions/950087/how-do-i-include-a-javascript-file-in-another-javascript-file

Then it exposes `window.io` and here we go.

Kind of defeats the purpose of stencil and bundling, but works for now.

**Comment by gabouh** [comment]
> I came up with a workaround to load `socketio-client` dynamically (not to bundle it at all), like this: https://stackoverflow.com/questions/950087/how-do-i-include-a-javascript-file-in-another-javascript-file
> 
> Then it exposes `window.io` and here we go.
> 
> Kind of defeats the purpose of stencil and bundling, but works for now.

Could you elaborate on how you had success with dynamic loading of the library?

**Comment by kirillgroshkov** [comment]
Approach similar to this:
```
function dynamicallyLoadScript(url) {
    var script = document.createElement("script");  // create a script DOM node
    script.src = url;  // set its src to the provided URL

    document.head.appendChild(script);  // add it to the end of the head section of the page (could change 'head' to 'body' to add it to the end of the body section instead)
}
```

Loading from e.g unpkg.com

**Comment by danawoodman** [comment]
I too am unable to get this working

```json
// package.json
"socket.io-client": "2.2.0"
```

```ts
import io from 'socket.io-client'
```

```
[ ERROR ]  Rollup: Missing Export: src/components/app-home/app-home.js:2:7
           'default' is not exported by
           ../node_modules/socket.io-client/lib/index.js

      L1:  import io from 'socket.io-client';
      L2:  export class AppHome {

[08:00.8]  rebuild failed, watching for
           changes... in 1.45 s
```

**Comment by m1nka** [comment]
I'm also having the issue that I cannot get stenciljs to work with socket.io-client. Any more suggestions how to solve this elegantly?

**Comment by bitflower** [comment]
Hey guys, I think the secret is a Stencil config like this:
```
import { Config } from '@stencil/core';
//import { sass } from '@stencil/sass';
import builtins from 'rollup-plugin-node-builtins';

export const config: Config = {
...
  plugins: [
    //sass(),
    builtins()
  ],
  nodeResolve: {
    browser: true,
    preferBuiltins: true // Workaround for https://github.com/ionic-team/stencil/issues/1326
  },
...
};
```

**Comment by OgheneMichael** [comment]
@bitflower 

The Stencil config fixed it for me... 

**Thank you**

**Comment by AnonymousArthur** [comment]
@bitflower 
This one works.
Thank you!

**Comment by jpchip** [comment]
The fix from @bitflower doesn't work in Stencil One (1.0.0-beta.2).

Using:
```
// package.json
"rollup-plugin-node-builtins": "2.1.2",
"socket.io-client": "2.2.0"

// app.tsx
import io from 'socket.io-client';
```

And the config from bitflower's post above, I get this error on `npm start`:

```
[ ERROR ]  Rollup: Invalid External Id
           'C:\workspace\my-components\node_modules\@stencil\core\dist\sys\src\empty.js' is imported as an
           external by
            C:\workspace\my-components\node_modules\@stencil\core\dist\sys\src\empty.js?commonjs-proxy, but is
           already an existing non-external module id.

[ WARN  ]  Bundling Warning UNRESOLVED_IMPORT
           'C:\workspace\my-components\node_modules\@stencil\core\dist\sys\src\empty.js' is imported by
            C:\workspace\my-components\node_modules\@stencil\core\dist\sys\src\empty.js?commonjs-proxy, but could
           not be resolved – treating it as an external dependency
```

Any thoughts?

**Comment by simonhaenisch** [comment]
@jpchip #1551 reports the same kind of error... I think it might be a bug.

**Comment by jpchip** [comment]
Hey, Stencil 1.0.0 still seems to have this issue (it works in 0.18.1-0). Any possible work arounds y'all can think of? I would love to upgrade to 1.0.0 but this is blocking me. I can throw together a repo with just this problem if that helps.

**Comment by jpchip** [comment]
I created a project recreating the specific problem, posting here so things don't get lost:

https://github.com/jpchip/stencil-socketio

**Comment by jpchip** [comment]
This issue is resolved in stencil v1.0.1! Thanks @manucorporat 

https://github.com/ionic-team/stencil/commit/8b4597ca28f50c4a3eb197ecb85fa34d17965f4c
